src/GitHub.InlineReviews/Services/PullRequestSessionService.cs

```diff
@@ -0,0 +1,133 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Threading.Tasks;
+using GitHub.Models;
+using GitHub.Services;
+using LibGit2Sharp;
+
+namespace GitHub.InlineReviews.Services
+{
+    /// <summary>
+    /// Provides a common interface for services required by <see cref="PullRequestSession"/>.
+    /// </summary>
+    [Export(typeof(IPullRequestSessionService))]
+    class PullRequestSessionService : IPullRequestSessionService
+    {
+        readonly IGitService gitService;
+        readonly IGitClient gitClient;
+        readonly IDiffService diffService;
+
+        readonly IDictionary<Tuple<string, string>, string> mergeBaseCache;
+
+        [ImportingConstructor]
+        public PullRequestSessionService(
+            IGitService gitService,
+            IGitClient gitClient,
+            IDiffService diffService)
+        {
+            this.gitService = gitService;
+            this.gitClient = gitClient;
+            this.diffService = diffService;
+
+            mergeBaseCache = new Dictionary<Tuple<string, string>, string>();
+        }
+
+        /// <inheritdoc/>
+        public async Task<IList<DiffChunk>> Diff(ILocalRepositoryModel repository, string baseSha, string relativePath, byte[] contents)
+        {
+            var repo = await GetRepository(repository);
+            return await diffService.Diff(repo, baseSha, relativePath, contents);
+        }
+
+        /// <inheritdoc/>
+        public async Task<string> GetTipSha(ILocalRepositoryModel repository)
+        {
+            var repo = await GetRepository(repository);
+            return repo.Head.Tip.Sha;
+        }
+
+        /// <inheritdoc/>
+        public async Task<bool> IsUnmodifiedAndPushed(ILocalRepositoryModel repository, string relativePath, byte[] contents)
+        {
+            var repo = await GetRepository(repository);
+
+            return !await gitClient.IsModified(repo, relativePath, contents) &&
+                   await gitClient.IsHeadPushed(repo);
+        }
+
+        public async Task<byte[]> ExtractFileFromGit(
+            ILocalRepositoryModel repository,
+            int pullRequestNumber,
+            string sha,
+            string relativePath)
+        {
+            var repo = await GetRepository(repository);
+
+            try
+            {
+                return await gitClient.ExtractFileBinary(repo, sha, relativePath);
+            }
+            catch (FileNotFoundException)
+            {
+                var pullHeadRef = $"refs/pull/{pullRequestNumber}/head";
+                await gitClient.Fetch(repo, "origin", sha, pullHeadRef);
```

Comment:
This needs to use an https remote, so it needs to do `var remote = await gitClient.GetHttpRemote(repo);` and call fetch with the returned remote.